-[Feature]: Multi-line input for adding bulk tracker URLs

mahee96 opened the pull request:
### Feature Request Reason:
- Adding Publicly available trackers in a bulk manner was not possible with the single entry texField in current addTracker AlertDialog.
- Helps a dead torrent when download speed has reduced to almost 0 and only leeches are available.
- Increasing the trackers by bulk adding popular trackers increases the chances of finding a peer who has the parts that is to required.
- Was very annoying not to have this feature since I loved iTorrent anyways :)

### Implementation Details:
- Created AlertDialog which has textView as containerView to receive multi-line inputs (trackers list with entries separated by new-line/carriage-return)

- Once the multi-line string is received, it is sanitized parsed and then used for adding the trackers as usual.
- Added output reporting for the trackers add-operation using alertDialog after processing add-request.

### Deliverables:
- updated TrackersListController and Dialog to add trackers from multiline text-input inside AlertDialog

### Edit: 1
- added `info` dialog for successful add operation.

### ScreenShots
<img src="https://github.com/XITRIX/iTorrent/assets/47920326/e3569eb7-e450-4907-bba9-1c3d591dc29f" width="230" height="500" />
<img src="https://github.com/XITRIX/iTorrent/assets/47920326/2e4a6ae4-a359-4a32-af37-654f1b87fcb4" width="230" height="500" />
<img src="https://github.com/XITRIX/iTorrent/assets/47920326/123f27bb-663b-4af4-9c57-29c6c5f3fc23" width="230" height="500" />
<img src="https://github.com/XITRIX/iTorrent/assets/47920326/7e69cd62-c96b-4761-b329-54924b4d2c0f" width="230" height="500" />
<img src="https://github.com/XITRIX/iTorrent/assets/47920326/fc9a027e-2309-43c8-b3e7-88bf45ce468a" width="230" height="500" />
<img src="https://github.com/XITRIX/iTorrent/assets/47920326/49d7daa9-6c95-45eb-a59e-5eea0043a963" width="230" height="500" />

mahee96 committed:
-[Feature]: Multi-line inputs for adding tracker URLs

- updated TrackersListController and Dialog to add trackers from multiline text-input inside AlertDialog

mahee96 commented:
@XITRIX , A small contribution from my side. Hoping to see this pull request getting merged and pushed in next release.

FYI, I hope the releases are still published to AltStore/Sidestore repositories.

Thanks.

XITRIX commented:
Hi, sorry for not responding for so long. Thanks for nice implementation, could you please move `EditTextView` and `EditTextViewController` in separate files somewhere in CustomViews folder. After that I'll be glad to merge this PR

mahee96 commented:
@XITRIX, sure no problem, I can do that. Since I am actively using iTorrent with this feature, I had noticed that the add logic to process all the trackers provided seems to run on main/UI thread and this is causing the renderer to show 9000+ms when adding huge list with more than 200+ trackers.

I will see how I can offload the task to background thread or at least keep it async from the UI thread or show progress bar when doing add on UI thread in sync mode.

But this nuisance should not stop from using the feature as all the 200+ trackers still get added.

Also the implications of having 200+ active trackers and its hit on performance if any needs to be profiled/tested, before fully rolling out.

mahee96 committed:
-[Feature]: Multi-line inputs for adding tracker URLs

- moved classes EditTextView and EditTextViewController into CustomViews as swift files

mahee96 committed:
-[Feature]: Multi-line inputs for adding tracker URLs

- Fix: updated add action to show processed valid text lines instead of including empty lines
- Feature: Added Info dialog upon successful add, to show processed entries vs valid trackers

mahee96 commented:
@XITRIX  Moved the UI View and its controller for EditTextView into CustomViews directory as requested. Please do the needful.

![image](https://github.com/XITRIX/iTorrent/assets/47920326/383bcbcd-edb4-45cf-89eb-ce7cb0b9a83a)

mahee96 commented:
> @XITRIX, sure no problem, I can do that. Since I am actively using iTorrent with this feature, I had noticed that the add logic to process all the trackers provided seems to run on main/UI thread and this is causing the renderer to show 9000+ms when adding huge list with more than 200+ trackers.
> 
> I will see how I can offload the task to background thread or at least keep it async from the UI thread or show progress bar when doing add on UI thread in sync mode.
> 
> But this nuisance should not stop from using the feature as all the 200+ trackers still get added.
> 
> Also the implications of having 200+ active trackers and its hit on performance if any needs to be profiled/tested, before fully rolling out.

~~@XITRIX , For this nuisance issue, I **will create a separate PR later** as it is **not a show stopper** to use this feature. 
If the user really wants a long list of trackers (200+) to be added in a bulk, we will inquire more diagnostics info from them if at all the UI delay occurs in such cases.~~

Edit 1: 
Please find my new commits and comment below on the update, No need for new PR, background processing is now included.

Thanks again.

mahee96 committed:
-[Feature]: Multi-line inputs for adding tracker URLs

- Sub-Feature: Moved parse and add logic to `(async)background thread`
- Sub-Feature: Added `Progress indicator` to show progress when performing actions in the background
    - involves using GrandCentralDispatch(GCD) to offload bulk of processing to bg thread and post progress updates on mainThread.
- Sub-Feature: Added distinctions between duplicates/invalid entries for the result dialog.
- Fix: updated add action result dialog to show valid text lines count in the `processed` field instead of all lines(including empty line inputs).

mahee96 committed:
-[Feature]: Multi-line inputs for adding tracker URLs

- Fix: updated url validator to accept more tracker url formats and other supported tracker url schemes.

mahee96 committed:
-[Fix]: Fixed typo from `shure` to `sure` in TrackersListController.

mahee96 commented:
@XITRIX, I have created new commits to offload the bulk of parsing and add logic to background/async. Have also tested it on simulator for inputs more than 25K lines (512 unique trackers).
The newly added progress circle also works as expected which works without blocking the UI Thread and `doesn't cause fence hang` anymore.
  
  
I understand that these tests aren't actual use case, since adding more than 150+ trackers is going to hog the router with those many TCP connections and basically slow down the overall download due to contention, but this huge of input was to show that the logic is scalable and can be used in other places in the framework if required.

### Test Inputs - tracker entries
[trackers_test_512_entries.txt](https://github.com/XITRIX/iTorrent/files/13045830/trackers_test_512_entries.txt)
[trackers_test_28K_entries.txt](https://github.com/XITRIX/iTorrent/files/13045828/trackers_test_28K_entries.txt)

![image](https://github.com/XITRIX/iTorrent/assets/47920326/756d78ed-8692-43ae-9f7b-e82b951af902)

> @XITRIX, sure no problem, I can do that. Since I am actively using iTorrent with this feature, I had noticed that the add logic to process all the trackers provided seems to run on main/UI thread and this is causing the renderer to show 9000+ms when adding huge list with more than 200+ trackers.
> 
> I will see how I can offload the task to background thread or at least keep it async from the UI thread or show progress bar when doing add on UI thread in sync mode.
> 
> But this nuisance should not stop from using the feature as all the 200+ trackers still get added.
> 
> Also the implications of having 200+ active trackers and its hit on performance if any needs to be profiled/tested, before fully rolling out.

XITRIX commented on iTorrent/ViewControllers/TorrentDetails/TrackersList/TrackersListController.swift:
```diff
@@ -112,23 +112,147 @@ class TrackersListController: ThemedUIViewController {
 
     @IBAction func removeAction(_ sender: UIBarButtonItem) {
-        let controller = ThemedUIAlertController(title: nil, message: NSLocalizedString("Are you shure to remove this trackers?", comment: ""), preferredStyle: .actionSheet)
+        let controller = ThemedUIAlertController(title: nil, message: NSLocalizedString("Are you sure to remove this trackers?", comment: ""), preferredStyle: .actionSheet)
```
Could you please update localization keys as well? (Or just leave key as it is and just fix the localization itself) Otherwise it will break other localizations

mahee96 replied:
@XITRIX, I have made the required changes in new commit e9a4d3e

Thanks

XITRIX commented:
Sorry for not updating for a long time, I've added 1 comment, could you please fix it and I'll merge this brunch

Thanks in advance

mahee96 committed:
-[Feature]: Multi-line input for adding bulk tracker URLs

- Fix: Synchronized localization changes across all Localization strings.

mahee96 commented:
I have updated the localizations in e9a4d3e, Thanks!

XITRIX commented:
Thanks a lot for your contribution!
